### DISCUSSION TITLE
Project files should be human readable

### DISCUSSION THREAD

**Discussion by mansellan** [twinbasic/twinbasic#131] [General]
**Is your feature request related to a problem?**
Modern source control systems expect to deal with non-binary files. Their diff tools are set up to optimise for this convention.

**Describe the solution you'd like**
Either JSON, XML, or (even) INI project files.

**Additional context**
Currently, twinBASIC uses:
1. workspace files (json)
2. twinproj files (binary, with source code embedded (as UTF-8?))

.Net (with the .Core reset) has moved to minimal, convention-based project files, expressed in XML form. This decision made it much easier to diff commits and critique any changes made (as there was much less boilerplate).

VB6, for all its faults with indeterministic serialisation, tended to keep with renderable file formats, except for binary resource files. Everything except .**x files could be diffed for source control.

Using a binary project file seems like a step backwards.

**Comment by WaynePhillipsEA** [comment]
You have to consider the twinproj file as a virtual file system.  It's like a zip file.

There are good reasons that the twinproj file format is not text based, in the same way that there are good reasons ZIP files, VDI files etc are not text based.  

The decision mostly comes down to efficiency, and serialization simplicity.  There is also meta-data that does not lend itself well to text-based storage, such as indices.  Soon, you'll also be able to include binary files such as graphics inside the twinproj file once the one-file limit has been removed.

When thinking of SCC, you wouldn't diff a ZIP file, and so you wouldn't diff a twinproj file.  You'd diff the extracted content, and once we support SCC from inside VS code, that's exactly how it'll work with tB projects.

If the twinproj file wasn't a self-contained virtual file system, then a text-based human readable file format would likely have been chosen.

**Reply by roke-julian-lockhart** [reply]
SCC?

**Reply by mansellan** [reply]
Source Code Control.

**Comment by mika76** [comment]
I must say I agree with the original poster, dealing with binary files in source control is a pain in the bum. There's nothing more frustrating that .frx files changing and you have no idea why. 

I also think it's pretty standard these days for projects files, and many many other files developers have to work with to be humanly readable, but even more importantly to be comparable and easily differenced.

What about when you have to do merges between people adding new files to a project, or as you say images but with the same file names?

People are even choosing to work with svg files more because of these features.

**Reply by WaynePhillipsEA** [reply]
You've missed the main point of my response.   The twinproj file is not like any ordinary 'project file', it is a *container*.  The elements within it *are* text files.  As such, you will never want to do SCC on the twinproj file itself, but instead do the source code/diffs on the elements within the **container**.  Again, think of it like a ZIP file container.

We will be providing a simple extraction tool for twinproj files, so that you can do SCC outside of VS code if you wish, but ultimately the main aim is to offer SCC *inside* VS code (which doesn't actually see the twinproj file itself, only the contents via its virtual file system).

The closest thing to a traditional 'project file' in twinBASIC is the Settings file contained within the twinproj file system.  That Settings file is a JSON text file, human readable, as you would expect.

**Reply by mika76** [reply]
I played with it a bit now, and I do understand what you mean, but I'm struggling to figure out why it's like this? 

What is the purpose of it being in a container, and not just a normal folder/folders with files? 

>The decision mostly comes down to efficiency, and serialization simplicity. There is also meta-data that does not lend itself well to text-based storage, such as indices. Soon, you'll also be able to include binary files such as graphics inside the twinproj file once the one-file limit has been removed.

I see you wrote the above, but it just seems to be adding complexity for very little gain:

* There's nothing simpler than serializing/de-serializing single files (which you'd still have to do in a virtual file system anyway), and meta data (including indices) can be kept in xml/json/base64 whatever is required.
* What if the file is corrupt, even with a single byte? Does that mean you've lost ALL your source code?
* Currently git/tfs/svn only sees the .twinproj file, so I assume you'd have to hack vscode to "see" the files in the virtual system so that you can use the tools to check in and out? What if someone wants to use a source control which is not standard, like mercurial or bazaar for example? Or even a different editor (like as mentioned in #82)
 
I personally feel that the industry has moved away from using binary files many years ago (for source assets) and probably for a good reason.

**Reply by WaynePhillipsEA** [reply]
> What is the purpose of it being in a container, and not just a normal folder/folders with files?

The virtual file system used by twinBASIC is designed for versatility.  As one example, the future plan is to be able to support in-memory VBA streams (where the underlying storage is based on [IStorage](https://docs.microsoft.com/en-us/windows/win32/api/objidl/nn-objidl-istorage)), and the virtual file system design is PERFECT for this.  The design of our virtual file system also pretty much matches the [VS code API virtual file system](https://code.visualstudio.com/api/references/vscode-api#FileSystem).  

> I see you wrote the above, but it just seems to be adding complexity for very little gain...

You are not privy to the intricate details of the compiler, and our future plans, so I can't imagine how you could possibly come to such a conclusion. 
 
One of the key features of the twinproj file, is that it allows for embedded files (as they are now), AND links to external files (not yet available in the preview).  For example, VBP files will automatically be migrated to a twinproj with just links to the external files, so the source code in _that_ case is not stored inside the twinproj file at all.

> Currently git/tfs/svn only sees the .twinproj file, so I assume you'd have to hack vscode to "see" the files in the virtual system so that you can use the tools to check in and out

Nope, there's no 'hack' required at all.  VS code fully supports virtual file systems, and indeed if you look at the recent VS code releases, a lot development work has been done in VS Code in support of the virtual file system approach.

> What if someone wants to use a source control which is not standard, like mercurial or bazaar for example?

If VS Code doesn't support it, then neither will twinBASIC.   You'll still be able to use the raw tools (coming soon) to extract from, and repopulate twinproj files, and then you can ultimately still make it work with any SCC you like.

The compiler has the virtual file system at its very core, and for very good reason.  This isn't changing any time soon.

**Reply by bclothier** [reply]
Can I ask about how we will cope with corruption? Currently with a VBA project, if it's corrupted, Microsoft gives us 2 choices: 1) make a backup of the corrupted file or 2) throw it out. 😨 In this case, the option 1 isn't a great option because this requires spelunking around in binary contents to find and fix the corruption. It's usually "cheaper" (in terms of time and effort) to go back to the backup and redo the work. 

Will it be able to recover as much as it can or at least make it possible for us to fix the corruption without having to revert to a prior backup? 

One more thing that is not quite clear to me. Since the twinproj file won't be in the SCC, from where does VSC? tB compiler? has the information to rebuild the twinproj file from scratch? Imagine a scenario where I'm cloning a repository on another machine and I don't have the twinproj file but want to reconstitute the project from the SCC.

**Reply by WaynePhillipsEA** [reply]
> Can I ask about how we will cope with corruption?

At the moment, the twinproj will be parsed as much as possible (i.e. 'recovery'), but obviously the success depends on many factors, and we don't want to have to rely on that.   It is planned that we offer an auto-backup feature for the twinproj file.

> from where does VSC? tB compiler? has the information to rebuild the twinproj file from scratch?

tB will be able to reconstruct the twinproj file fully from the repository.   At the moment, things like breakpoints are stored as meta data rather than actual files in the virtual filesystem, but this will be changing soon so that literally everything will be able to be cloned to/from a repository.

FWIW, VS code just introduced a feature that creates a virtual file system for accessing github repositories 'live'.  So by using a virtual file system we are very much on-board with where VS code is heading, IMO.

**Reply by dmrvb** [reply]
I also much prefer files in plain text (NOT XML) but if binary is essential then a mixed format file is usually preferable to separate files. Trouble with mixed format, especially if its "mostly text" is that sooner or later somebody (probably me) will try to edit it.

I have needed to create my own mixed format file definitions for a couple of my VB6 projects. I have two rules:
The first line or two of the file **must** be text and describe the file format, something like:

File Format = TwinBasic Project
File Version = 1.0 

Any binary sections must be enclosed in a block and the btye size specified, something like:

Binary Bytes =100
Begin Block = Binary
////binary data here////
End Block = Binary

This makes it relatively easy to inspect the file (in Notepad) and to write code to extract data from the file whilst skipping over the binary blocks if they are not of interest.

**Comment by DavidHollman** [comment]
It sounds like the twinproj file is almost like a local cache, rather than the 'master record'. (Except for project metadata which doesn't seem to have any text-only equivalent that would be able to be source controlled?) So why make the twinproj file user-visible at all... if its just needed for efficiency. This would be something like the .vs directory which Visual Studio creates when you work on a C# or other .NET solution, but ultimately can be totally discarded and rebuilt when needed, the user doesn't concern themselves with it.

**Reply by fafalone** [reply]
This thread is a little outdated as it's discussing the Visual Studio Code addin, what tB was before it had it's own IDE. Now, .twinproj files are the entire project; there's no other file besides external references like TLB refs/OCX controls. You can export it to individual files, but those are a copy of source files rather than what you're working with in the IDE.

**Reply by wqweto** [reply]
I tend to image of .twinproj files as .mdb files (because of Wayne's background) but its a actually virtual file-system so its more like a .zip file I guess.

Don't like it as I'm pro github diff tracking but waiting to find a reason about this early decision at some point. This or it being reverted to separate source files referenced from a .twinproj file like in good old .vbp times :-))

**Reply by fafalone** [reply]
Wayne has said that eventually there will be optional support for using a .twinproj file just like a vbp, where the source files are separate. Which is what I prefer too because I like to use common modules shared between projects for certain things; keeping them separate lets them be edited from any project using it and then other projects don't need to update their own copy to get the changes.

**Reply by dmrvb** [reply]
Yes, I work in exactly the same way.

**Reply by DavidHollman** [reply]
I use diffing and source control as a fundamental work process. Files like FRX or MDBs which don't fit that model are much harder to work with. We've actually taken pains to eliminate binary files in as many places as possible so that work can be more productive and consistent. Nearly all popular programming languages historically have used a text-only approach, VB6 being the unfortunate exception with FRX's. 

It would be ideal for us if tB would stick to the norm of easily diffable text-only files. Having to do a manual import/export procedure to work with tB would produce some disincentives for people to do things the "right" way (in terms of our normal process). 

So indeed an approach where .twin files function more like .vbp files would be much appreciated! :)   In that case I suggest there might be two distinct project "formats" with different names / file extensions. One analogous to VBP (or CSPROJ, VBPROJ, etc.) and a second where everything is all bundled together.

**Reply by WaynePhillipsEA** [reply]
As mentioned above, this is already planned.

For the time being, please accept the current situation and allow focus to continue on the UI side of things.

**Reply by DavidHollman** [reply]
@WaynePhillipsEA sorry if my comment seemed pushy, that was not at all how I intended it!  

I don't think this is a significant issue at all at the present time, I'm sure your current focus areas are spot on. I just wanted to register my personal opinion for the long haul.

**Reply by WaynePhillipsEA** [reply]
No worries @DaveInCaz, understood :)  No apology necessary, it's all good.

**Reply by gh658804** [reply]
Hi @WaynePhillipsEA. Please clarify what is planned (whenever it is planned, it is understandable that it's not a priority at this particular moment). Is it this? ->

> It sounds like the twinproj file is **almost like a local cache**, rather than the 'master record'. (Except for project metadata which doesn't seem to have any text-only equivalent that would be able to be source controlled?) **So why make the twinproj file user-visible at all**... if its just needed for efficiency. **This would be something like the .vs directory which Visual Studio creates when you work on a C# or other .NET solution, but ultimately can be totally discarded and rebuilt when needed**, the user doesn't concern themselves with it.

If the answer is Yes then great. Having plain text files (like i.e. .bas.twb or cls.twb) would indeed be great, and minimize the number of binary files as much as possible. IMO, VB's .frx files could have been text files from the beginning.

**Comment by rexxitall** [comment]
Fully agree ! Would make it way more easy to migrate things between vb6 vba and TB. As a first step it would be a help if the file extnesions would be same as in Vb6. To just name all .twin is unhandy and you can not just throw files around. Also The Module and Class statemenst should be hidden. Its unhandy by copy and paste a whole module. Even if the project file isnt VB6 compatible. its easy to create just a tb project export ist and create a new VB6 project with the exported files imported. In this case you has what you want (at least half the way) - one code base (tb) but you can see without a lot of work how the "original" would behave. Which might also a nice timesaver to track problems down ;)

**Reply by WaynePhillipsEA** [reply]
> Fully agree ! Would make it way more easy to migrate things between vb6 vba and TB. As a first step it would be a help if the file extnesions would be same as in Vb6. To just name all .twin is unhandy and you can not just throw files around. Also The Module and Class statemenst should be hidden. Its unhandy by copy and paste a whole module. Even if the project file isnt VB6 compatible. its easy to create just a tb project export ist and create a new VB6 project with the exported files imported. In this case you has what you want (at least half the way) - one code base (tb) but you can see without a lot of work how the "original" would behave. Which might also a nice timesaver to track problems down ;)

Nope, none of that is happening, sorry, and it is non-negotiable.

**We support legacy BAS/CLS files 100%.  You can continue to use them without issue.**

TWIN files have important differences.  The main one is *full Unicode* support.  BAS/CLS files are ANSI based (based on your system ANSI codepage).  They cannot support Unicode, and any attempt to make them Unicode-friendly will break backwards compatibility with VBx that does not interpret them as Unicode.   The second difference is that they were designed to allow for multiple components per file.

So as you can see, it is very important to offer the new format, and it is imperative be able to distinguish between legacy vs new format via the file extension.  These decisions weren't made lightly or without significant thought.

**Reply by rexxitall** [reply]
Do as you want as usual. User needs are how to say ... just hindering :D

**Reply by mansellan** [reply]
Uh, wut? You had a comprehensive answer. It's not possible.

**Comment by Greedquest** [comment]
This discussion seems like an X/Y issue. Most of the desire to have human readable files seems to be around git and source control. I'm sure you've considered this @WaynePhillipsEA  so can I ask what is your latest vision for source control support? I think if that is clear then most of this discussion is moot. 

Re-reading the other posts in this discussion I see 3 options (not necessarily exclusive, multiple alternative options could be supported at once if there is no single "Right Way"):

 1. Export on save dumps a copy of the live project in human readable format to the Disk. Sync with folder could bring changes from the human readable format back into the editor. This is the workflow RubberDuck allows us to do. It is annoying as if I make changes to the exported human readable files these are not automatically reflected in the live project open in the VBE (VBA/VB6 Editor). 
      - RD could add a filesystem-watcher to spot changes in the exported files and automatically sync to the VBE, this is what Notepad++ does, but there is some risk of cache invalidation.
 3. VBP style "symbolic links" to source code. This creates a live mirror between the project in the virtual filesystem in the IDE and the real filesystem. Make an edit in one place and it is immediately reflected in the other. No risk of cache invalidation if the Operating System manages it for you.
 4. git integrated into the IDE, which means the files do not need to be exported since git would work within the virtual filesystem. This makes sense for tB as a VS Code extension as it has good support. However, I worry that unlike VSCode, there won't be resources available to polish user interface for the git integration to the same standard in tB's IDE. I know RD gave up on git integration for this reason, deemed beyond the scope and not bringing much value that wasn't already possible with code sync + external git tools. I like to use a 3rd party GUI [GitKraken](https://www.gitkraken.com/)
rather than command line git for source control, so I prefer editors that support method 2.


**Additional Context:**
If tB exposes some tokeniser/abstract syntax tree/reflection API, it will give the community a lot more power than was previously possible to create automatic code linting and formatting tools. Auto-formatters, indenters, code to re-order your `Subs` in alphabetical order, etc.

These could be worked in as IDE plugins if the IDE supports scripting. Or as Language Servers. Both those methods work with a virtual filesystem (LSP is `FileSystemProvider` agnostic, the LSP Server that provides the linting or indenting etc. has no idea where the file comes from, as this info is hidden behind abstraction and the responsibility of the LSP tool - i.e. the IDE). That said my favourite and simplest to author way is to create exes that run as git hooks/[pre-commit hooks](https://pre-commit.com/) - e.g. run the smart-indenter every time I try to make a git commit. Issue: these git-hooks cannot work with a virtual filesystem.

**Reply by Greedquest** [reply]
Sounds like the vision is all 3 (or maybe not having IDE git integration working off the virtual filesystem)

**Reply by bclothier** [reply]
Just to throw this out. I _think_ it's theoretically possible to write a shell extension handler to treat a `.twinproj` as a folder, allowing you to browse inside the `.twinproj` just like you can do inside a `.zip` folder without uncompressing it. If it can be mounted via Windows explorer, then I think the concerns should go away. If this is in fact possible, then that can be a community project. Would be neat to have a shell extension handler written in tB to show how easy it is to customize Windows explorer. :)

**Reply by DavidHollman** [reply]
I'm not sure its exactly an X/Y problem... the filesystem is a really good abstraction layer for integrating different tools - including future uses which can't be expected right now. Its a reliable lowest common denominator. There are so many ways in which manipulating plain files can be an advantage.

The filesystem also works well as a cross-platform abstraction; Windows shell plugins obviously would not though maybe something equivalent exists on a Mac or Linux distros.

Git is not universal ... that's probably especially true in corporate environments where devs may have less choice. (My group happens to use Mercurial which we did choose ourselves... but before GitHub became overwhelmingly popular.)

We happen to do mixed development in VB6 and C# (not to mention XAML, SQL, JSON, etc.), source control all of that together, and do benefit from having a common process that works for everything. But maybe that amount of mixed tech is not typical.

**Reply by fafalone** [reply]
@bclothier yes that's possible. 

But when there's not even physical files behind it, it's *not* going to be easy or simple. Such a shell extension would be quite complex, especially if you to want to support properties.

**Reply by mika76** [reply]
Also a shell extension is windows only, while this project aims to be cross-platform in the future...

**Reply by fafalone** [reply]
I'm sure there's equivalents for other platforms. You wouldn't be able to make a single solution for all platforms for this; it's too low level. 

A quick search shows you could create a virtual file system for both Linux and MacOS... Is the plan to make the IDE itself work on Android, or just to compile for it? Android doesn't like letting you do anything low level so that might be the only problem.

**Reply by Greedquest** [reply]
I don't really know how the topic of shell extensions has come up. That's not really what I'm getting at - in fact it's the X of the X/Y I'm referring to. Having a shell extension is a user interface thing,  it won't impact the ability of a source control tool to track the internal files of a project.

@DaveInCaz Fair enough, I should have said Source Control not just Git. IIUC whilst VSCode has a minimal git GUI built in, other source control tools can interact as plugins through the Language Server Protocol (LSP) - this again does not care if it's a virtual filesystem + .twinproj, or a real filesystem + .vbp symbolic links. The drawback is you'd have to write the plugin. The LSP is versatile enough that all source control integration including git could be done as LSP Servers (additional apps that plug into tB IDE). Most probably already exist for VSCode since the LSP is editor-agnostic.  However the real filesystem as you say is the most flexible since any source control tool, like pre-commit hooks, GitKraken, TortoiseGit, etc. can work with the filesystem but only a subset make sense as LSP servers.

**Reply by fafalone** [reply]
It could allow 3rd party source control tools that don't understand the .twinproj format to track the internal files by being able to access them as a part of the file system, rather than manually parsing the twinproj file itself.

**Reply by Greedquest** [reply]
@fafalone I thought a shell extension was more along the lines of how I can open a zip file in Windows explorer as if it is a regular folder. But if I try to commit the zip file to a repo with git, it still sees it as a big binary blob. That is not diffable. So I still don't quite follow how a shell extension can help existing source control tools without them having special treatment in the source control software? Or do the source control tools also need modification to make this work?

**Reply by fafalone** [reply]
It depends on how the shell extension is implemented. 

The zip folder extension doesn't implement all features of a normal folder, and is in fact quite a bit different from a traditional shell extension; but you could. 

For instance if you attempt to drag a file from zip folder onto GitHub, it won't allow you too. This is because it doesn't supply a CF_HDROP clipboard format; you could have the shell extension extract to a temporary physical location and supply a CF_HDROP format to it, then you'd be able to drop it on GitHub's website directly from the virtual folder; this would be transparent to the user.

**Reply by Greedquest** [reply]
@fafalone ok shell extensions are more powerful than I thought. Still, do you know of any flags that you could in theory support which would allow git to track changes to the contents of the twinproj file? Without dragging and dropping to GitHub to commit the files.

**Reply by fafalone** [reply]
I don't know how git works under the hood but if you skip the shell and just go low level APIs to copy files, no it won't be possible as you can only present it as a folder to higher level methods. But the point is it allows *some* options for browsing the files on disk and using them with other tools, as opposed to none when there's no way of browsing it as a folder.

**Comment by rexxitall** [comment]
I still think to implement something compatible to VBx would be the simplest and most flexible solution. Its also the most robust ones.  If something weired happens you do not loose everything :) And a lot of code tools alredy written will have a chance to still work on :) There are reasons why all compilers i know relay on simple ascii files ;) Some kind of packing system might be a idea to archive things. But that would be on top.

**Reply by roke-julian-lockhart** [reply]
@rexxitall, you already got an answer to this at https://github.com/twinbasic/twinbasic/discussions/131#discussioncomment-4106116

**Comment by melanysoft** [comment]
Hi everyone, I've just joined the community, truly amazed with what Master Wayne has achieved so far. We're still using VB6 on a daily basis in my company, sharing our work between a dozen developers. I've been keeping twinBASIC on my radar for quite some time now. There are many great things about it, and a lot more to come, but the topic of code versioning was one of my biggest concerns. In my environment, we are occasionally working on the same project, unlocking only files we need to maintain or improve. We're forking, merging, and doing everything a team does in any programming environment, not just VB. Being able to do so is crucial for building and maintaining large and complex projects.

Presuming twinBASIC is aiming at big projects, I strongly suggest making it as editable in text editors as possible.

As I see it, *.twin virtual file system should be treated as something local, even temporary. As someone pointed out, like the .vs cache folder. This file will likely be on the ignore list of any SCC system. What matters is the plain text representation of the project; in the latest version, it is automatically saved with the project file.

Also, a big thing in VB6 (or any other IDE I work with) is the possibility to group several projects and debug them together. This is essential when writing libraries (COM or plain). Is that supported in twinBASIC already, or will it be ?

I couldn't find the latest thoughts on this topic, but this would be my wish whishlist  :)

1. Open the twinBASIC project from the plain text project file (yes, pretty much like .sln or .vbp).
2. If opening from a plain text project file, twinBasic should automatically refresh the *.twin "cache".
4. We should be able to put different projects in groups and run/debug them at once.

Best regards

**Reply by WaynePhillipsEA** [reply]
Firstly thanks for your kind words - much appreciated.  I'm glad to hear that you're looking at twinBASIC.

The plan is that the twinproj file will become equivalent to, or at least can be used much like, a VBP project file.   Source files will then be linked to, rather than embedded into, the twinproj container.  

Those currently working with SCC alongside twinBASIC projects typically use the 'Export' feature - usually the automatic 'Export on save' option, which exports all the raw text files from the contained twinproj file into a typical folder hierarchy.  This works somewhat, but of course is quite limited and you can't share code simply between projects.

Whilst project groups are not currently supported, you can potentially use packages in the meantime.   You can compile each project into a package, and reference that package from another project (or nested package).  Doing so allows segregation of the projects, but most importantly the compiler allows you to debug and step-into the code inside referenced packages.

To answer your questions specifically;

1) Yes, raw source files will be supported soon.
2) Yes, if we do use a cache/history inside the twinproj container, it will be automatically kept in-sync
3) This is absolutely planned, but support for this will happen in the v1-v2 release cycle.  For now, packages might be an interim solution.

**Reply by melanysoft** [reply]
Thank you for such a quick reply, Wayne. I can imagine the amount of workload you're handling at the moment.

Your reply put a smile on my face. It's good to know this issue is somewhere on the roadmap, so we will eventually get there.

Didn't know about the packages stuff. Thanks for pointing it out. It's pretty much how Java does the thing, so I guess it's "read-only" when debugging the package code. I'll give it a try.

Best regards

**Reply by fafalone** [reply]
Correct, packages are read-only when loaded in another project. But you could keep the source project for the package open in another instance of tB, so once the debugger points you where to go, you can quickly modify then reload the package-- which can be done from a local twinpack file that's generated when you build a package, in addition to private and public online package servers. To update a package, first uncheck it in References, and when asked, click 'Remove it' to delete it from the file system. Then click on the Available Packages' tab and either check it on the server list, or use 'Import from file' to load a local twinpack (note: you must be on the packages tab otherwise the import file dialog will only load COM/ActiveX references), and then tick the box for it.

**Reply by mansellan** [reply]
Would also just add that binary packages will also be supported in the future, for library vendors that don't wish to make their source code available.

**Reply by mansellan** [reply]
When external files are a thing, the question arises of how the project settings should be exposed...

1. Ini files? I mean, yuck, but it could be made directly compatible with .vbp...
2. YAML: Concise, but... problematic for something so complex.
3. JSON: Presumably close to the existing representation, might be the easiest way?
4. XML: Kinda old skool.
5. Something else? I'm sure other languages use other project formats, thoughts?

**Reply by melanysoft** [reply]
> Correct, packages are read-only when loaded in another project. ... click 'Remove it' to delete it from the file system...

Amazing! I guess at that moment the code execution must be stopped, right ?
Can't imagine how you could make it resume the debug session...

Thank you for sharing this great approach.

**Reply by fafalone** [reply]
> > Correct, packages are read-only when loaded in another project. ... click 'Remove it' to delete it from the file system...
> 
> Amazing! I guess at that moment the code execution must be stopped, right ? Can't imagine how you could make it resume the debug session...
> 
> Thank you for sharing this great approach.

Yes sorry if I wasn't clear, you'd have to exit your app and return to design mode to update a package, it's similar to a regular reference-- even after tB supports Edit and Continue I don't think you could hot swap a package. (E&C is coming, there were some early demos a while back but it never hit production ready, yet 🙂)

**Reply by Greedquest** [reply]
> Would also just add that binary packages will also be supported in the future, for library vendors that don't wish to make their source code available.

ActiveX dll?

**Reply by fafalone** [reply]
Who wouldn't want to avoid the baggage of AX registration hell or SxS slightly-less-hellish-hell?

**Reply by bclothier** [reply]
A HCKU Ax would be less problematic as that can be much more easily updated even in most locked-down environment. 

That said, AX means only strict COM compatibility. For example, overloads wouldn't be exposed. Maybe in future there'll be a binary object file to allow for static linking of the packages without the source code.

**Comment by dmrvb** [comment]
I would also like an option for individual files, and ideally human readable.
I maintain three medium/large projects. Each has its own files in a number of folders, but all share a lot of files stored in a set of "common" folders. When I update any common code then that is saved back to the common folders to then be available in all the projects. I am not sure how to easily do this in tB at present.
For various reasons I do like to look at files in a simple text editor (and other utilities) from time to time.

**Comment by rexxitall** [comment]
[profanity removed].  My suggestion to enhance it would be a 4096 AES full encryption as well. Than they are in every case gone ... *zynism off* Even Microsoft was not that stupid in the old days. At least you got usually something to recover. Efficency is also a joke on modern procesors. How often they are read ? Thousends of times or more or less just once ? Next stupid question. How can someone make tools for them as addon ? Reverse engineering or fat fee and NDA ? So plain text for sure. Well documented. It can not be that hard to write a simple parser for it. If you got no idea, just ask a Ai. Usually she will return you a working boilerplate.

**Reply by WaynePhillipsEA** [reply]
No swearing.  2nd warning.
